### PR TITLE
[SPARK-5323][SQL] Remove Row's Seq inheritance.

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

**Comment by rxin** [comment]
cc @yhuai @marmbrus @liancheng you should probably all take a look at this since it changes so many things.

Our test cases are really hard to change :(

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala [hunk 1]
```diff
@@ -154,8 +154,16 @@ class JoinedRow extends Row {
   }
 
   override def toString() = {
-    val row = (if (row1 != null) row1 else Seq[Any]()) ++ (if (row2 != null) row2 else Seq[Any]())
-    s"[${row.mkString(",")}]"
+    // Make sure toString never throws NullPointerException.
+    if ((row1 eq null) && (row2 eq null)) {
+      "[ empty row ]"
+    } else if (row1 eq null) {
+      "[" + row2.mkString(",") + "]"
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: Maybe this `row2.mkString("[", ",", "]")`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala [hunk 2]
```diff
@@ -197,45 +205,45 @@ class JoinedRow2 extends Row {
     this
   }
 
-  def iterator = row1.iterator ++ row2.iterator
+  override def toSeq: Seq[Any] = row1.toSeq ++ row2.toSeq
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: This can be much more costly then `iterator`. However, I wonder whether we use it anywhere.

File: sql/core/src/main/scala/org/apache/spark/sql/columnar/InMemoryColumnarTableScan.scala [hunk 3]
```diff
@@ -128,8 +127,7 @@ private[sql] case class InMemoryRelation(
             rowCount += 1
           }
 
-          val stats = Row.fromSeq(
-            columnBuilders.map(_.columnStats.collectedStatistics).foldLeft(Seq.empty[Any])(_ ++ _))
+          val stats = Row.merge(columnBuilders.map(_.columnStats.collectedStatistics) : _*)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/columnar/InMemoryColumnarTableScan.scala]: Remove the space before `:`? (Not very sure about the convention here.)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala [hunk 4]
```diff
@@ -144,7 +144,7 @@ package object debug {
       case (null, _) =>
 
       case (row: Row, StructType(fields)) =>
-        row.zip(fields.map(_.dataType)).foreach { case(d,t) => typeCheck(d,t) }
+        row.toSeq.zip(fields.map(_.dataType)).foreach { case(d,t) => typeCheck(d,t) }
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala]: Space after `,`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala [hunk 5]
```diff
@@ -291,12 +309,61 @@ trait Row extends Seq[Any] with Serializable {
...
+    }
+    MurmurHash3.finalizeHash(h, n)
+  }
+
+  /* ---------------------- utility methods for Scala ---------------------- */
+
+  /**
+   * Return a Scala Seq representing the row. ELements are placed in the same order in the Seq.
+   */
+  def toSeq: Seq[Any]
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala]: Would like to have a `toArray` here, since most `Row` subclasses are backed by array, which already has plenty rich operations.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala [hunk 6]
```diff
@@ -66,7 +66,7 @@ private[sql] object CatalystConverter {
 
   // TODO: consider using Array[T] for arrays to avoid boxing of primitive types
   type ArrayScalaType[T] = Seq[T]
-  type StructScalaType[T] = Seq[T]
+  type StructScalaType[T] = Row
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala]: I'm a bit surprised that this actually compiles...

File: sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala [hunk 7]
```diff
@@ -227,30 +234,30 @@ class ParquetFilterSuite extends QueryTest with ParquetTest {
...
+      checkFilterPushdown1(rdd, '_1)(Literal("2") >   '_1, classOf[Lt  [java.lang.String]])(Seq(Row("1")))
+      checkFilterPushdown1(rdd, '_1)(Literal("3") <   '_1, classOf[Gt  [java.lang.String]])(Seq(Row("4")))
+      checkFilterPushdown1(rdd, '_1)(Literal("1") >=  '_1, classOf[LtEq[java.lang.String]])(Seq(Row("1")))
+      checkFilterPushdown1(rdd, '_1)(Literal("4") <=  '_1, classOf[GtEq[java.lang.String]])(Seq(Row("4")))
 
-      checkFilterPushdown(rdd, '_1)(!('_1 < "4"), classOf[Operators.GtEq[java.lang.String]])("4")
-      checkFilterPushdown(rdd, '_1)('_1 > "2" && '_1 < "4", classOf[Operators.And])("3")
-      checkFilterPushdown(rdd, '_1)('_1 < "2" || '_1 > "3", classOf[Operators.Or]) {
+      checkFilterPushdown1(rdd, '_1)(!('_1 < "4"), classOf[Operators.GtEq[java.lang.String]])(Seq(Row("4")))
+      checkFilterPushdown1(rdd, '_1)('_1 > "2" && '_1 < "4", classOf[Operators.And])(Seq(Row("3")))
```
Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala]: `checkFilterPushdown` is enough for these checks, no need to wrap them with `Seq(Row(...))`.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala (hunk 6 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala]: Row used to be a Seq so why not?

File: sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala (hunk 7 above)
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala]: no that one takes only an int?

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala]: it is not possible to override it because we used fancy currying. if we had used just a normal argument then I could've overriden it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala (hunk 5 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala]: it's not that bad in that case since it just returns a wrappedarray. we can do it in a separate pr if we think it is important.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/WrapDynamic.scala [hunk 8]
```diff
@@ -39,7 +39,8 @@ case class WrapDynamic(children: Seq[Attribute]) extends Expression {
   override def eval(input: Row): DynamicRow = input match {
     // Avoid copy for generic rows.
     case g: GenericRow => new DynamicRow(children, g.values)
-    case otherRowType => new DynamicRow(children, otherRowType.toArray)
+    case otherRowType => new DynamicRow(children, otherRowType.toSeq.toArray)
+    // TODO: toSeq.toArray is very slow!
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/WrapDynamic.scala]: We might even just get rid of WrapDynamic

File: sql/core/src/main/scala/org/apache/spark/sql/execution/pythonUdfs.scala [hunk 9]
```diff
@@ -143,7 +143,8 @@ object EvaluatePython {
    * Convert Row into Java Array (for pickled into Python)
    */
   def rowToArray(row: Row, fields: Seq[DataType]): Array[Any] = {
-    row.zip(fields).map {case (obj, dt) => toJava(obj, dt)}.toArray
+    // TODO: this is slow!
+    row.toSeq.zip(fields).map {case (obj, dt) => toJava(obj, dt)}.toArray
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/pythonUdfs.scala]: Haha, in this case probably not slower than python...

File: sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD.scala [hunk 10]
```diff
@@ -474,6 +474,16 @@ private[sql] object JsonRDD extends Logging {
             valWriter(field.dataType, v)
         }
         gen.writeEndObject()
+
+      case (StructType(ty), v: Row) =>
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD.scala]: Can the case above be removed now?

File: sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala (hunk 7 above)
Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala]: But why did you make `checkFilterPushdown` only accept an `Int`? The original version accepts `Any`.

Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetFilterSuite.scala]: Yeah, I admit currying is not necessary and troublesome here... I can fix it in another PR.
